### PR TITLE
Activiti/Activiti#3845 add permission property when getting task by Id

### PR DESCRIPTION
**Pull request by Giovanni007**
Ref: https://github.com/Activiti/Activiti/issues/3845

### CODE REVIEW

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            if(!canUserViewTask(task)) return;
+            
+            permissions.add(TaskPermissions.VIEW);
+            
+            if(canUserClaimTask(task, userId, userGroups)) {
+                permissions.add(TaskPermissions.CLAIM);
+            }
+            
+            if(isUserAssignee(task, userId)) {
+                permissions.add(TaskPermissions.RELEASE);
```
Comment (balsarori) [inline activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java:51]: `RELEASE` should be only possible if the task is assigned to current user and the user is a candidate or a member of a candidate group

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    
+    private boolean canUserViewTask(TaskEntity task) {
+        return taskControllerHelper.canUserViewTask(QTaskEntity.taskEntity.id.eq(task.getId()));
+    }
+    
+    private boolean canUserClaimTask(TaskEntity task, String userId, List<String> userGroups) {
+        boolean isUserOwner = task.getOwner() != null && task.getOwner().equals(userId);
+        
+        return !isTaskAssigned(task) && (isUserOwner || isUserCandidate(task, userId, userGroups));
```
Comment (balsarori) [inline activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java]: task owner can't  claim a task only candidate users or members of candidate groups can claim a task

**Review by balsarori — CHANGES_REQUESTED** [review]

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java (hunk 1 above)
Comment (Giovanni007) [reply activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java:51]: I think is not necessary to check if the user is also a candidate of a member if he is the assignee. I also found this in rb [TaskRuntimeImpl.java#L475-L485](https://github.com/Activiti/Activiti/blob/ecec52d763b4ce0fe0362a0f3996b28bd877c318/activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/impl/TaskRuntimeImpl.java#L475-L485).
 What do you think?

Comment (balsarori) [reply activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java:51]: My understanding to `RELEASE` is to release back the task to the queue. So if there is no candidates or candidate groups where can this task release back to?

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java (hunk 2 above)
Comment (Giovanni007) [reply activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java]: In my understanding, the owner is the creator of a standalone task, and he can claim a task also if he is not a candidate user or member of a candidate group.

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java (hunk 1 above)
Comment (Giovanni007) [reply activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java:51]: I'm not sure this scenario is possible. Can we replicate it? I guess that If you are the assignee, you were a candidate user or member of a candidate group.  I think the permissions should follow the behaviour of the rb API.

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java (hunk 2 above)
Comment (balsarori) [reply activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java]: Standalone tasks aren't claimable.

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java (hunk 1 above)
Comment (balsarori) [reply activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java:51]: `RELEASE` is opposite to `CLAIM` so if the task wasn't claimed it shouldn't be possible to release it.

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java (hunk 2 above)
Comment (balsarori) [reply activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java]: Owner can reassign the task (that's how it works in APS)

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            if(canUserClaimTask(task, userId, userGroups)) {
+                permissions.add(TaskPermissions.CLAIM);
+            }
+            
+            if(canUserReleaseTask(task, userId, userGroups)) {
+                permissions.add(TaskPermissions.RELEASE);
+            }
+            
+            if(isUserAssignee(task, userId)) {
+                permissions.add(TaskPermissions.UPDATE);
```
Comment (balsarori) [inline activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java:55]: Can you give owner update permissions too? By default current user is [set as owner when creating a new standalone task](https://github.com/Activiti/Activiti/blob/develop/activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/impl/TaskRuntimeImpl.java#L251) therefore owner should be able to update it later.

File: activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    private boolean canUserViewTask(TaskEntity task) {
+        return taskControllerHelper.canUserViewTask(QTaskEntity.taskEntity.id.eq(task.getId()));
+    }
+    
+    private boolean canUserClaimTask(TaskEntity task, String userId, List<String> userGroups) {
+        return !isTaskAssigned(task) && isUserCandidate(task, userId, userGroups);
+    }
+
+    private boolean canUserReleaseTask(TaskEntity task, String userId, List<String> userGroups) {
+        return isUserAssignee(task,userId) && isUserCandidate(task, userId, userGroups);
```
Comment (balsarori) [inline activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java]: I think it should be enough to check that the task has candidate users and/or groups. Current user might get removed from candidates list via REST API. My previous concern is that task shouldn't be released if there are no candidates.

**Review by balsarori — APPROVED** [review]

**Review by mauriziovitale — APPROVED** [review]

**Pull request merged**